// errors.js/errors.js
define("errors", ["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    class ErrorList {
        constructor() {
            this.es = new Array();
        }
        handleError(err) {
            this.es.push(err.Err());
        }
        handleErrorList(el) {
            el.ForEach((err) => this.es.push(err));
        }
        handleErrorArray(ea) {
            ea.forEach((err) => this.handleError(err));
        }
        Err() {
            return this.es.join(", ");
        }
        ForEach(fn) {
            this.es.forEach(fn);
        }
        Push(err) {
            switch (err.constructor) {
                case ErrorList:
                    // Handle as an error list
                    return this.handleErrorList(err);
                case Array:
                    // Handle as an array of generic Errors
                    return this.handleErrorArray(err);
                default:
                    // Handle as generic Error
                    return this.handleError(err);
            }
        }
    }
    exports.ErrorList = ErrorList;
    class ErrorString {
        constructor(err) {
            this.err = err;
        }
        Err() {
            return this.err;
        }
    }
    exports.ErrorString = ErrorString;
});
